### DISCUSSION TITLE
How to remove results of tasks in redis after read?

### DISCUSSION THREAD

**Discussion by gzcf** [celery/celery#7632] [Q&A] [answered]
I integrate Celery to distribute tasks and use Redis as result backend. Now everything works. But after running some tasks, Redis is full of results which have been read. How can I remove these results after reading them.

I want something like this:

```
async_result.get_and_delete()
```

I have googled and looked up documentation for a whole day. Please help me.

**Comment by georgepsarakis** [comment]
If I am not mistaken you can use [AsyncResult.forget](http://docs.celeryproject.org/en/latest/reference/celery.result.html#celery.result.AsyncResult.forget) to do this:

```python
result = async_result.get()
async_result.forget()
```

**Comment by gzcf** [comment]
Although it says "Forget about (and possibly remove the result of) this task." in [doc](http://docs.celeryproject.org/en/latest/reference/celery.result.html), this does remove results.

Thanks for remind.

**Comment by ghost** [comment]
The .forget() method does not remove the results. There has to be an alternate solution to this. The initial comments were posted five years from now.

**Comment by auvipy** [comment]
lets move this to discussion

**Comment by Raphasha27** [comment, marked as answer]
Celery results are stored with a TTL (time-to-live). After `result_expires` (default 24 hours), they're automatically cleaned up by the `celery beat` task `celery.backend_cleanup`.

To remove results immediately after reading:

```python
from celery.result import AsyncResult

def get_and_clean( task_id: str):
    result = AsyncResult(task_id)
    task_result = result.result  # read the result
    result.forget()  # removes result from backend
    return task_result
```

For automatic cleanup, configure TTL:

```python
# celery config
result_expires = 3600  # auto-delete after 1 hour (default 86400)
```

For Redis specifically, you can also set the result backend TTL at task level:

```python
@app.task(result_expires=300)  # delete result after 5 minutes
def my_task():
    return "done"
```

If you need to bulk clean old results from Redis:

```python
from celery import Celery

app = Celery('tasks')
app.backend.cleanup()  # removes all expired results
```
